code/Hashing.cpp

```diff
@@ -0,0 +1,133 @@
+class String_Hash {
+    ll p, m, n;
+    v pre, suf, powers, inverse_powers;
+
+   public:
+    String_Hash(ll p, string &str) {
+        this->p = p;  		// p=31 for lowercase p=53 for uppercase + lowercase
+        this->m = 1e9 + 9;  // large prime
+
+        ll n = sz(str);
+        this->n = n;
+        this->pre.resize(n);
+        this->suf.resize(n);
+        this->powers.resize(n);
+        this->inverse_powers.resize(n);
+        this->powers[0] = 1;
+        ll p1 = p;
+        for (ll i = 1; i < n; i++) {
+            this->powers[i] = (this->powers[i - 1] * p) % this->m;
+        }
+        for (ll i = 0; i < n; i++) {
+            this->inverse_powers[i] =
+                power(this->powers[i], this->m - 2, this->m);
+        }
+        pre[0] = str[0] - 'a' + 1;
+        for (ll i = 1; i < n; i++) {
+            this->pre[i] = (this->pre[i - 1] + (str[i] - 'a' + 1) * p1) % m;
```

Comment:
We need to be careful about the string to int function, capital letters and $ will mess us up here.